@claude two things:
- address CodeRabbit review comments
- fix failed unitessts:
```
FAILED tests/unit/test_runners.py::TestWorkspaceRouting::test_first_matching_workspace_wins - AssertionError: 'default' != 'ws-a'
- default
+ ws-a
FAILED tests/unit/test_runners.py::TestWorkspaceRouting::test_route_applied_when_input_matches_pattern - AssertionError: 'default' != 'routed-ws'
- default
+ routed-ws
FAILED tests/unit/test_runners.py::TestWorkspaceRouting::test_route_overrides_default_workspace - AssertionError: 'default' != 'routed-ws'
- default
+ routed-ws
FAILED tests/unit/test_runners.py::TestWorkspaceRouting::test_wildcard_pattern_matches_subdomain - AssertionError: 'default' != 'vulnweb-ws'
- default
+ vulnweb-ws
======================== 4 failed, 805 passed in 53.78s ========================
```

--- a/secator/runners/_base.py
+++ b/secator/runners/_base.py
@@ -198,7 +198,7 @@ class Runner:
 		# Determine inputs
 		self.debug(f'resolving inputs with {len(self.dynamic_opts)} dynamic opts', obj=self.dynamic_opts, sub='init')
 		self.inputs = [inputs] if not isinstance(inputs, list) else inputs
-		self.inputs = list(set(self.inputs))
+		self.inputs = list(dict.fromkeys(self.inputs))
 		if self.caller != 'Task':
 			targets = [Target(name=target) for target in self.inputs]
 			for target in targets:
@@ -215,6 +215,16 @@ class Runner:
 		profiles_str = run_opts.get('profiles') or []
 		self.debug('resolving profiles', obj={'profiles': profiles_str}, sub='init')
 		self.profiles = self.resolve_profiles(profiles_str)
+
+		# Apply route-based workspace if no profile/explicit workspace was set
+		default_ws = CONFIG.workspace.default or 'default'
+		if not self.workspace_explicit and self.workspace_name == default_ws:
+			route_workspace = self._resolve_route_workspace(self.inputs)
+			if route_workspace:
+				self.debug(f'route workspace -> {route_workspace}', sub='init')
+				self.workspace_name = route_workspace
+				self.context['workspace_name'] = route_workspace
+				self.context['workspace_id'] = route_workspace
 
 		# Determine exporters
 		exporters_str = self.run_opts.get('output') or self.default_exporters
@@ -1438,14 +1448,6 @@ class Runner:
 			self.workspace_name = profile_workspace
 			self.context['workspace_name'] = profile_workspace
 			self.context['workspace_id'] = profile_workspace
-		elif not self.workspace_explicit and self.workspace_name == default_ws:
-			# No profile workspace and no explicit -ws: apply route-based workspace if routes match
-			route_workspace = self._resolve_route_workspace(self.inputs)
-			if route_workspace:
-				self.debug(f'route workspace -> {route_workspace}', sub='init')
-				self.workspace_name = route_workspace
-				self.context['workspace_name'] = route_workspace
-				self.context['workspace_id'] = route_workspace
 
 		# Apply exporters (consumed from run_opts['output'] right after profile resolution)
 		if profile_exporters is not None:

--- a/tests/unit/test_config.py
+++ b/tests/unit/test_config.py
@@ -259,6 +259,32 @@ class TestConfig(unittest.TestCase):
 		config2 = Config.parse(path=self.config_test)
 		self.assertEqual(config2.workspace.routes['my_ws'], ['*vulnweb.com*', '*ocervell*'])
 
+	def test_workspace_routes_remove_existing_pattern(self):
+		"""Removing an existing pattern leaves the remaining patterns intact."""
+		from secator.config import Config
+		config = Config.parse(path=self.config_test)
+		config.set('workspace.routes.my_ws', '*vulnweb.com*', strategy='append')
+		config.set('workspace.routes.my_ws', '*ocervell*', strategy='append')
+		config.set('workspace.routes.my_ws', '*vulnweb.com*', strategy='remove')
+		self.assertEqual(config.workspace.routes['my_ws'], ['*ocervell*'])
+
+	def test_workspace_routes_remove_missing_pattern(self):
+		"""Removing a non-existent pattern is a no-op (warns but does not raise)."""
+		from secator.config import Config
+		config = Config.parse(path=self.config_test)
+		config.set('workspace.routes.my_ws', '*ocervell*', strategy='append')
+		config.set('workspace.routes.my_ws', '*doesnotexist*', strategy='remove')
+		self.assertEqual(config.workspace.routes['my_ws'], ['*ocervell*'])
+
+	def test_workspace_routes_remove_workspace_key(self):
+		"""Unsetting a workspace key (no value) deletes it from routes entirely."""
+		from secator.config import Config
+		config = Config.parse(path=self.config_test)
+		config.set('workspace.routes.my_ws', '*ocervell*', strategy='append')
+		self.assertIn('my_ws', config.workspace.routes)
+		config.unset('workspace.routes.my_ws')
+		self.assertNotIn('my_ws', config.workspace.routes)
+
 
 @mock.patch('sys.stderr', devnull)
 class TestConfigEnv(unittest.TestCase):

--- a/tests/unit/test_runners.py
+++ b/tests/unit/test_runners.py
@@ -740,46 +740,73 @@ class TestCommandRunner(unittest.TestCase):
 class TestWorkspaceRouting(unittest.TestCase):
 	"""Tests for automatic workspace routing based on input targets."""
 
+	def setUp(self):
+		from secator.utils_test import clear_modules
+		clear_modules()
+		from secator.runners import Command
+		from secator.utils_test import mock_command
+
+		class LocalMyCommand(Command):
+			input_types = ['slug']
+			cmd = 'dummy'
+			input_flag = '-u'
+			file_flag = None
+
+		self.Cmd = LocalMyCommand
+		self.mock_command = mock_command
+
 	def test_route_applied_when_input_matches_pattern(self):
 		"""Workspace route is applied when an input matches a configured pattern."""
+		import unittest.mock
 		routes = {'routed-ws': ['*host1*']}
-		with patch('secator.config.CONFIG.workspace.routes', new=routes), \
-			patch('secator.config.CONFIG.workspace.default', new=''):
-			with mock_command(MyCommand, ['host1'], {}, []) as cmd:
+		with unittest.mock.patch('secator.runners.task.Task.get_task_class', return_value=self.Cmd), \
+			patch('secator.runners._base.CONFIG.workspace.routes', new=routes), \
+			patch('secator.runners._base.CONFIG.workspace.default', new=''), \
+			patch('secator.runners._base.CONFIG.profiles.defaults', []):
+			with self.mock_command(self.Cmd, ['host1'], {}, []) as cmd:
 				self.assertEqual(cmd.workspace_name, 'routed-ws')
 				self.assertEqual(cmd.context.get('workspace_name'), 'routed-ws')
 				self.assertEqual(cmd.context.get('workspace_id'), 'routed-ws')
 
 	def test_route_not_applied_when_no_input_matches(self):
 		"""Workspace route is NOT applied when no input matches any pattern."""
+		import unittest.mock
 		routes = {'routed-ws': ['*vulnweb.com*']}
-		default = ''
-		with patch('secator.config.CONFIG.workspace.routes', new=routes), \
-			patch('secator.config.CONFIG.workspace.default', new=default):
-			with mock_command(MyCommand, ['secator.cloud'], {}, []) as cmd:
+		with unittest.mock.patch('secator.runners.task.Task.get_task_class', return_value=self.Cmd), \
+			patch('secator.runners._base.CONFIG.workspace.routes', new=routes), \
+			patch('secator.runners._base.CONFIG.workspace.default', new=''), \
+			patch('secator.runners._base.CONFIG.profiles.defaults', []):
+			with self.mock_command(self.Cmd, ['secator.cloud'], {}, []) as cmd:
 				self.assertEqual(cmd.workspace_name, 'default')
 
 	def test_route_not_applied_when_workspace_explicit(self):
 		"""Routes are NOT applied when user explicitly passes -ws, even when value equals the default."""
+		import unittest.mock
 		routes = {'routed-ws': ['*host1*']}
 		# Set default to 'my-default'; user explicitly passes the same value.
 		# Without workspace_explicit=True, routes would match and override — but with it, they should not.
-		with patch('secator.config.CONFIG.workspace.routes', new=routes), \
-			patch('secator.config.CONFIG.workspace.default', new='my-default'):
+		with unittest.mock.patch('secator.runners.task.Task.get_task_class', return_value=self.Cmd), \
+			patch('secator.runners._base.CONFIG.workspace.routes', new=routes), \
+			patch('secator.runners._base.CONFIG.workspace.default', new='my-default'), \
+			patch('secator.runners._base.CONFIG.profiles.defaults', []):
 			opts = {'context': {'workspace_name': 'my-default', 'workspace_explicit': True}}
-			with mock_command(MyCommand, ['host1'], opts, []) as cmd:
+			with self.mock_command(self.Cmd, ['host1'], opts, []) as cmd:
 				self.assertEqual(cmd.workspace_name, 'my-default')
 
 	def test_route_overrides_default_workspace(self):
 		"""Workspace route overrides the configured default when no explicit -ws passed."""
+		import unittest.mock
 		routes = {'routed-ws': ['*host1*']}
-		with patch('secator.config.CONFIG.workspace.routes', new=routes), \
-			patch('secator.config.CONFIG.workspace.default', new='my-default'):
-			with mock_command(MyCommand, ['host1'], {}, []) as cmd:
+		with unittest.mock.patch('secator.runners.task.Task.get_task_class', return_value=self.Cmd), \
+			patch('secator.runners._base.CONFIG.workspace.routes', new=routes), \
+			patch('secator.runners._base.CONFIG.workspace.default', new='my-default'), \
+			patch('secator.runners._base.CONFIG.profiles.defaults', []):
+			with self.mock_command(self.Cmd, ['host1'], {}, []) as cmd:
 				self.assertEqual(cmd.workspace_name, 'routed-ws')
 
 	def test_enforced_profile_overrides_route(self):
 		"""An enforced profile workspace overrides a route-matched workspace."""
+		import unittest.mock
 		from secator.template import TemplateLoader
 		routes = {'routed-ws': ['*host1*']}
 		profile = TemplateLoader(input={
@@ -788,14 +815,17 @@ class TestWorkspaceRouting(unittest.TestCase):
 			'enforce': True,
 			'workspace': 'enforced-ws',
 		})
-		with patch('secator.config.CONFIG.workspace.routes', new=routes), \
-			patch('secator.config.CONFIG.workspace.default', new=''):
+		with unittest.mock.patch('secator.runners.task.Task.get_task_class', return_value=self.Cmd), \
+			patch('secator.runners._base.CONFIG.workspace.routes', new=routes), \
+			patch('secator.runners._base.CONFIG.workspace.default', new=''), \
+			patch('secator.runners._base.CONFIG.profiles.defaults', []):
 			opts = {'profiles': [profile]}
-			with mock_command(MyCommand, ['host1'], opts, []) as cmd:
+			with self.mock_command(self.Cmd, ['host1'], opts, []) as cmd:
 				self.assertEqual(cmd.workspace_name, 'enforced-ws')
 
 	def test_non_enforced_profile_overrides_route(self):
 		"""A non-enforced profile workspace takes precedence over a route match."""
+		import unittest.mock
 		from secator.template import TemplateLoader
 		routes = {'routed-ws': ['*host1*']}
 		profile = TemplateLoader(input={
@@ -803,26 +833,34 @@ class TestWorkspaceRouting(unittest.TestCase):
 			'type': 'profile',
 			'workspace': 'profile-ws',
 		})
-		with patch('secator.config.CONFIG.workspace.routes', new=routes), \
-			patch('secator.config.CONFIG.workspace.default', new=''):
+		with unittest.mock.patch('secator.runners.task.Task.get_task_class', return_value=self.Cmd), \
+			patch('secator.runners._base.CONFIG.workspace.routes', new=routes), \
+			patch('secator.runners._base.CONFIG.workspace.default', new=''), \
+			patch('secator.runners._base.CONFIG.profiles.defaults', []):
 			opts = {'profiles': [profile]}
-			with mock_command(MyCommand, ['host1'], opts, []) as cmd:
+			with self.mock_command(self.Cmd, ['host1'], opts, []) as cmd:
 				self.assertEqual(cmd.workspace_name, 'profile-ws')
 
 	def test_wildcard_pattern_matches_subdomain(self):
 		"""A wildcard pattern like *vulnweb.com* matches subdomains."""
+		import unittest.mock
 		routes = {'vulnweb-ws': ['*vulnweb.com*']}
-		with patch('secator.config.CONFIG.workspace.routes', new=routes), \
-			patch('secator.config.CONFIG.workspace.default', new=''):
-			with mock_command(MyCommand, ['testphp.vulnweb.com'], {}, []) as cmd:
+		with unittest.mock.patch('secator.runners.task.Task.get_task_class', return_value=self.Cmd), \
+			patch('secator.runners._base.CONFIG.workspace.routes', new=routes), \
+			patch('secator.runners._base.CONFIG.workspace.default', new=''), \
+			patch('secator.runners._base.CONFIG.profiles.defaults', []):
+			with self.mock_command(self.Cmd, ['testphp.vulnweb.com'], {}, []) as cmd:
 				self.assertEqual(cmd.workspace_name, 'vulnweb-ws')
 
 	def test_first_matching_workspace_wins(self):
 		"""When multiple workspaces have patterns, the first matching workspace is used."""
+		import unittest.mock
 		routes = {'ws-a': ['*host1*'], 'ws-b': ['*host*']}
-		with patch('secator.config.CONFIG.workspace.routes', new=routes), \
-			patch('secator.config.CONFIG.workspace.default', new=''):
-			with mock_command(MyCommand, ['host1'], {}, []) as cmd:
+		with unittest.mock.patch('secator.runners.task.Task.get_task_class', return_value=self.Cmd), \
+			patch('secator.runners._base.CONFIG.workspace.routes', new=routes), \
+			patch('secator.runners._base.CONFIG.workspace.default', new=''), \
+			patch('secator.runners._base.CONFIG.profiles.defaults', []):
+			with self.mock_command(self.Cmd, ['host1'], {}, []) as cmd:
 				self.assertEqual(cmd.workspace_name, 'ws-a')
 
 

